File: backend/routers/fpo.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
import logging

from backend.database import get_db
from backend.models import FPO, Farmer
from backend.schemas import FPOCreate, FPOUpdate, FPOResponse

router = APIRouter()
logger = logging.getLogger(__name__)


@router.post("/", response_model=FPOResponse, status_code=status.HTTP_201_CREATED)
def create_fpo(fpo: FPOCreate, db: Session = Depends(get_db)):
    """Create a new Farmer Producer Organization"""
    try:
        # Check if registration number already exists
        existing_fpo = db.query(FPO).filter(FPO.registration_number == fpo.registration_number).first()
        if existing_fpo:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Registration number already exists"
            )
        
        db_fpo = FPO(**fpo.model_dump())
        db.add(db_fpo)
        db.commit()
        db.refresh(db_fpo)
        
        logger.info(f"Created FPO: {db_fpo.id}")
        return db_fpo
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error creating FPO: {str(e)}")
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create FPO"
        )


@router.get("/", response_model=List[FPOResponse])
def list_fpos(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    """List all FPOs with pagination"""
    try:
        fpos = db.query(FPO).offset(skip).limit(limit).all()
        return fpos
    except Exception as e:
        logger.error(f"Error listing FPOs: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve FPOs"
        )


@router.get("/{fpo_id}", response_model=FPOResponse)
def get_fpo(fpo_id: int, db: Session = Depends(get_db)):
    """Get a specific FPO by ID"""
    try:
        fpo = db.query(FPO).filter(FPO.id == fpo_id).first()
        if not fpo:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="FPO not found"
            )
        return fpo
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error retrieving FPO {fpo_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve FPO"
        )


@router.put("/{fpo_id}", response_model=FPOResponse)
def update_fpo(fpo_id: int, fpo_update: FPOUpdate, db: Session = Depends(get_db)):
    """Update an FPO's information"""
    try:
        db_fpo = db.query(FPO).filter(FPO.id == fpo_id).first()
        if not db_fpo:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="FPO not found"
            )
        
        # Update only provided fields
        update_data = fpo_update.model_dump(exclude_unset=True)
        for field, value in update_data.items():
            setattr(db_fpo, field, value)
        
        db.commit()
        db.refresh(db_fpo)
        
        logger.info(f"Updated FPO: {fpo_id}")
        return db_fpo
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating FPO {fpo_id}: {str(e)}")
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to update FPO"
        )


@router.delete("/{fpo_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_fpo(fpo_id: int, db: Session = Depends(get_db)):
    """Delete an FPO"""
    try:
        db_fpo = db.query(FPO).filter(FPO.id == fpo_id).first()
        if not db_fpo:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="FPO not found"
            )
        
        db.delete(db_fpo)
        db.commit()
        
        logger.info(f"Deleted FPO: {fpo_id}")
        return None
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting FPO {fpo_id}: {str(e)}")
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to delete FPO"
        )


@router.get("/{fpo_id}/members", response_model=List)
def get_fpo_members(fpo_id: int, db: Session = Depends(get_db)):
    """Get all farmers belonging to an FPO"""
    try:
        fpo = db.query(FPO).filter(FPO.id == fpo_id).first()
        if not fpo:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="FPO not found"
            )
        
        farmers = db.query(Farmer).filter(Farmer.fpo_id == fpo_id).all()
        return farmers
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error retrieving FPO members for {fpo_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve FPO members"
        )

File: src/routes/fpo.ts
import { Router, Request, Response } from "express";
import { db } from "../data.js";
import { FPOCreateInput, FPOUpdateInput } from "../types.js";

export const fpoRouter = Router();

// POST /api/v1/fpo - Create a new FPO
fpoRouter.post("/", (req: Request, res: Response) => {
  const { name, registration_number, address, contact_person, contact_email, contact_phone } = req.body as FPOCreateInput;

  if (!name || !registration_number) {
    return res.status(422).json({
      detail: [
        { loc: ["body", !name ? "name" : "registration_number"], msg: "Field required", type: "value_error.missing" }
      ]
    });
  }

  // Check if registration number already exists
  const existing = db.getFPOByRegistrationNumber(registration_number);
  if (existing) {
    return res.status(400).json({ detail: "Registration number already exists" });
  }

  const fpo = db.createFPO({
    name,
    registration_number,
    address: address ?? null,
    contact_person: contact_person ?? null,
    contact_email: contact_email ?? null,
    contact_phone: contact_phone ?? null,
  });

  return res.status(201).json(fpo);
});

// GET /api/v1/fpo - List all FPOs with pagination
fpoRouter.get("/", (req: Request, res: Response) => {
  const skip = parseInt(req.query.skip as string, 10) || 0;
  const limit = parseInt(req.query.limit as string, 10) || 100;

  const fpos = db.getFPOs(skip, limit);
  return res.json(fpos);
});

// GET /api/v1/fpo/:fpo_id - Get a specific FPO by ID
fpoRouter.get("/:fpo_id", (req: Request, res: Response) => {
  const fpoId = parseInt(req.params.fpo_id, 10);
  if (isNaN(fpoId)) {
    return res.status(404).json({ detail: "FPO not found" });
  }

  const fpo = db.getFPOById(fpoId);
  if (!fpo) {
    return res.status(404).json({ detail: "FPO not found" });
  }

  return res.json(fpo);
});

// PUT /api/v1/fpo/:fpo_id - Update an FPO's information
fpoRouter.put("/:fpo_id", (req: Request, res: Response) => {
  const fpoId = parseInt(req.params.fpo_id, 10);
  if (isNaN(fpoId)) {
    return res.status(404).json({ detail: "FPO not found" });
  }

  const existing = db.getFPOById(fpoId);
  if (!existing) {
    return res.status(404).json({ detail: "FPO not found" });
  }

  const updates = req.body as FPOUpdateInput;

  const updated = db.updateFPO(fpoId, {
    ...(updates.name !== undefined && { name: updates.name }),
    ...(updates.address !== undefined && { address: updates.address }),
    ...(updates.contact_person !== undefined && { contact_person: updates.contact_person }),
    ...(updates.contact_email !== undefined && { contact_email: updates.contact_email }),
    ...(updates.contact_phone !== undefined && { contact_phone: updates.contact_phone }),
    ...(updates.member_count !== undefined && { member_count: updates.member_count }),
  });

  return res.json(updated);
});

// DELETE /api/v1/fpo/:fpo_id - Delete an FPO
fpoRouter.delete("/:fpo_id", (req: Request, res: Response) => {
  const fpoId = parseInt(req.params.fpo_id, 10);
  if (isNaN(fpoId)) {
    return res.status(404).json({ detail: "FPO not found" });
  }

  const success = db.deleteFPO(fpoId);
  if (!success) {
    return res.status(404).json({ detail: "FPO not found" });
  }

  return res.status(204).send();
});

// GET /api/v1/fpo/:fpo_id/members - Get all farmers belonging to an FPO
fpoRouter.get("/:fpo_id/members", (req: Request, res: Response) => {
  const fpoId = parseInt(req.params.fpo_id, 10);
  if (isNaN(fpoId)) {
    return res.status(404).json({ detail: "FPO not found" });
  }

  const existing = db.getFPOById(fpoId);
  if (!existing) {
    return res.status(404).json({ detail: "FPO not found" });
  }

  const members = db.getFPOMembers(fpoId);
  return res.json(members);
});
